tmd/areas/create_area_weights.py

```diff
@@ -204,66 +220,79 @@ def create_area_weights_file(area: str, write_file: bool = True):
-    Return loss_function_value using the optimized weights and optionally
-    write the weights file.
+    Return target_loss_function_value using the optimized weights and
+    optionally write the weights file.
     """
+    print(f"CREATING WEIGHTS FILE FOR AREA {area} ...")
+    jax.config.update("jax_platform_name", "cpu")  # ignore GPU/TPU if present
+    jax.config.update("jax_enable_x64", True)  # use double precision floats
+
     # construct variable matrix and target array and weights_scale
     vdf = all_taxcalc_variables()
     target_matrix, target_array, weights_scale = prepared_data(area, vdf)
     wght_us = np.array(vdf.s006 * weights_scale)
     num_weights = len(wght_us)
     num_targets = len(target_array)
     print(f"USING {area}_targets.csv FILE CONTAINING {num_targets} TARGETS")
-    loss = loss_function_value(wght_us, target_matrix, target_array)
-    print(f"US_PROPORTIONALLY_SCALED_LOSS_FUNCTION_VALUE= {loss:.9e}")
+    loss = target_loss_function_value(wght_us, target_matrix, target_array)
+    print(f"US_PROPORTIONALLY_SCALED_TARGET_LOSS_FUNCTION_VALUE= {loss:.9e}")
     density = np.count_nonzero(target_matrix) / target_matrix.size
     print(f"target_matrix sparsity ratio = {(1.0 - density):.3f}")
 
-    # optimize weights by minimizing sum of squared wght*var-target deviations
-    if OPTIMIZE_RATIOS:
-        tar_matrix = (target_matrix * wght_us[:, np.newaxis]).T
-        lob = OPTIMIZE_LO_BOUND_RATIO * np.ones(num_weights)
-        hib = OPTIMIZE_HI_BOUND_RATIO * np.ones(num_weights)
-    else:  # if optimizing the weights
-        tar_matrix = target_matrix.T
-        lob = OPTIMIZE_LO_BOUND_RATIO * wght_us
-        hib = OPTIMIZE_HI_BOUND_RATIO * wght_us
+    # optimize weight ratios by minimizing the sum of squared deviations
+    # of area-to-us weight ratios from one such that the optimized ratios
+    # hit all of the area targets
+    #   NOTE: the optimization method is a version of the augmented
+    #         Lagrangian method that is simplified because of the
+    #         constraints (one for each target) are linear and
+    #         because the objective function is very simple (the
+    #         sum of the squared weight ratio deviations from one).
+    #         These simplifications obviate the need to iterate from
+    #         a lower constraint multiple to higher constraint multiple.
+    #         The multiple is the "mu sub k" variable in the third
+    #         equation in the "General method" section of the Wikipedia
+    #         article on the augmented Lagrangian method:
+    #         https://en.wikipedia.org/wiki/Augmented_Lagrangian_method
+    #         Instead of applying the multiple to the constraint functions,
+    #         here we apply the inverse of the multiple (which is called
+    #         the REGULARIZATION_LAMBDA) to the sum of the squared weight
+    #         ratio deviations from one, which is logically equivalent.
```

Comment:
I think documentation here is valuable. There are more direct analogies than the augmented Lagrangian. This is closely related to ridge regression and also to regularization in optimization (widely used in machine learning), particularly L2 regularization. Here's a suggested change that focuses more on regularization, while still mentioning the augmented Lagrangian.

```
# This optimization problem aims to find weight ratios x that minimize
# the sum of squared deviations from target values while keeping the
# weight ratios close to 1. The problem can be formulated as:

# Minimize: sum((Ax - b)^2) + lambda * sum((x - 1)^2)

# The first term represents the squared deviations between the matrix-vector
# product Ax and the target vector b, ensuring that the optimized ratios x
# meet the area targets (or other constraints represented by b). The second term
# is a regularization term that penalizes large deviations of the weight ratios x 
# from 1, where lambda controls the strength of this regularization.

# This is closely related to ridge regression, which penalizes large regression
# coefficients, except that here the regularization term is (x - 1) rather than x
# (we penalize differences from 1 rather than from 0). More generally, it is a form
# of L2 regularization widely used in machine learning. Good sources include
# Boyd, Stephen, and Lieven Vandenberghe. "Convex Optimization." Cambridge
# University Press, 2004, which is available for free at 
# https://web.stanford.edu/~boyd/cvxbook/bv_cvxbook.pdf. 
# Regularization is discussed in section 6.3.

# The approach also is related to the augmented Lagrangian method used for general 
# constraint optimization, but it simplifies the problem because the constraints 
# (Ax = b) are linear and the objective function is quadratic in nature. This 
# allows us to avoid using a Lagrange multiplier (the "mu sub k" in augmented 
# Lagrangian methods), and instead apply the inverse regularization lambda directly
# to the second term. Conceptually, this simplifies the problem while maintaining
# the logical equivalence to the regular augmented Lagrangian method.

# For more on the augmented Lagrangian method, see the general description here:
# https://en.wikipedia.org/wiki/Augmented_Lagrangian_method
```

Reply:
Note: No relation to the Boyd in the Convex Optimization book!